File: backend/app/discount.py
"""
Discount calculation — ported 1:1 from src/lib/serverprice/discount.ts so the
backend and the frontend's offline fallback always agree.
"""

from app.data import REGION_ADJUST, REGION_SHORT, SEGMENT_RULES
from app.models import (
    ConfidenceFactor,
    DiscountQuote,
    QuoteRequest,
    QuoteRule,
    ServerProduct,
)


def calculate_discount(product: ServerProduct, req: QuoteRequest) -> DiscountQuote:
    rules: list[QuoteRule] = []

    segment_rule = SEGMENT_RULES[req.segment]
    rules.append(
        QuoteRule(
            label=f"Segment baseline — {req.segment.replace('-', ' ')}",
            delta=segment_rule["base"],
            detail=segment_rule["note"],
        )
    )

    matching_tiers = [
        t
        for t in product.discountTiers
        if req.units >= t.minUnits and t.label.startswith(str(req.termMonths))
    ]
    tier = max(matching_tiers, key=lambda t: t.percent) if matching_tiers else None

    if tier:
        rules.append(
            QuoteRule(
                label=f"Commit tier — {tier.label}",
                delta=tier.percent,
                detail=f"Published tier for {tier.minUnits}+ units on this SKU",
            )
        )
    else:
        nearest = min(product.discountTiers, key=lambda t: t.minUnits, default=None)
        rules.append(
            QuoteRule(
                label="Commit tier — not reached",
                delta=0,
                detail=(
                    f"Needs {nearest.minUnits} units on a {nearest.label} to unlock {nearest.percent}%"
                    if nearest
                    else "No published commit tiers for this SKU"
                ),
            )
        )

    region = REGION_ADJUST[product.region]
    rules.append(
        QuoteRule(
            label=f"Region adjustment — {REGION_SHORT[product.region]}",
            delta=region["delta"],
            detail=region["reason"],
        )
    )

    if product.availability in ("constrained", "backorder"):
        penalty = -4 if product.availability == "backorder" else -2
        rules.append(
            QuoteRule(
                label="Supply constraint hold-back",
                delta=penalty,
                detail=f"{product.leadTimeDays}-day lead time — pricing desk restricts concessions",
            )
        )

    if req.units >= 50:
        rules.append(
            QuoteRule(
                label="Volume override",
                delta=3,
                detail=f"{req.units} units clears the 50-unit strategic volume gate",
            )
        )

    raw = sum(r.delta for r in rules)
    max_allowed = 100 - product.marginFloorPercent
    discount_percent = max(0, min(raw, max_allowed))

    if raw > max_allowed:
        rules.append(
            QuoteRule(
                label="Margin floor cap",
                delta=round(max_allowed - raw, 1),
                detail=f"Capped at {max_allowed}% to protect the {product.marginFloorPercent}% margin floor",
            )
        )

    list_total = product.basePrice * req.units * req.termMonths
    net_unit_price = product.basePrice * (1 - discount_percent / 100)
    net_total = net_unit_price * req.units * req.termMonths

    # Confidence is scored from real signals, and each signal that moves the
    # score records the reason it moved — the UI renders these verbatim rather
    # than guessing a canned explanation from the final band.
    score = 55
    factors: list[ConfidenceFactor] = []

    sample_size = segment_rule["sampleSize"]
    if sample_size > 200:
        score += 20
        factors.append(
            ConfidenceFactor(
                label="Deep deal history",
                detail=f"{sample_size} closed {req.segment.replace('-', ' ')} deals back this rate",
                direction="up",
            )
        )
    elif sample_size > 100:
        score += 12
        factors.append(
            ConfidenceFactor(
                label="Moderate deal history",
                detail=f"{sample_size} comparable {req.segment.replace('-', ' ')} deals on record",
                direction="up",
            )
        )
    else:
        score += 4
        factors.append(
            ConfidenceFactor(
                label="Thin deal history",
                detail=f"Only {sample_size} comparable {req.segment.replace('-', ' ')} deals to reference",
                direction="down",
            )
        )

    if tier:
        score += 14
        factors.append(
            ConfidenceFactor(
                label="Exact rate card match",
                detail=f"Hits the published {tier.label} tier at {tier.minUnits}+ units",
                direction="up",
            )
        )
    else:
        factors.append(
            ConfidenceFactor(
                label="No commit tier matched",
                detail="Priced off segment and region rules alone — no published tier applies",
                direction="down",
            )
        )

    if product.availability == "in-stock":
        score += 10
        factors.append(
            ConfidenceFactor(
                label="Supply confirmed",
                detail=f"In stock, {product.leadTimeDays}-day lead time",
                direction="up",
            )
        )
    elif product.availability == "backorder":
        score -= 12
        factors.append(
            ConfidenceFactor(
                label="On backorder",
                detail=f"{product.leadTimeDays}-day lead time puts the quote date at risk",
                direction="down",
            )
        )
    else:
        factors.append(
            ConfidenceFactor(
                label="Supply constrained",
                detail=f"{product.leadTimeDays}-day lead time — allocation not guaranteed",
                direction="down",
            )
        )

    margin_percent = 100 - product.marginFloorPercent - discount_percent
    if margin_percent < 5:
        score -= 18
        factors.append(
            ConfidenceFactor(
                label="Close to margin floor",
                detail=f"Only {round(margin_percent, 1)}% headroom above the {product.marginFloorPercent}% floor",
                direction="down",
            )
        )

    score = max(12, min(97, score))
    confidence = "high" if score >= 78 else "medium" if score >= 55 else "low"

    return DiscountQuote(
        productId=product.id,
        segment=req.segment,
        units=req.units,
        termMonths=req.termMonths,
        basePrice=product.basePrice,
        listTotal=list_total,
        discountPercent=round(discount_percent, 1),
        netUnitPrice=round(net_unit_price, 2),
        netTotal=round(net_total, 2),
        savings=round(list_total - net_total, 2),
        confidence=confidence,
        confidenceScore=score,
        marginPercent=round(margin_percent, 1),
        requiresApproval=discount_percent > 20 or margin_percent < 6,
        rules=rules,
        confidenceFactors=factors,
    )

File: src/lib/serverprice/discount.ts
import { REGION_SHORT, SEGMENT_RULES } from "./data";
import type { CustomerSegment, DiscountQuote, ServerProduct } from "./types";

const REGION_ADJUST: Record<string, { delta: number; reason: string }> = {
  "us-east": { delta: 0, reason: "Primary region — rate card baseline" },
  "us-west": { delta: -1, reason: "Capacity constrained region, less headroom" },
  "eu-central": { delta: 1.5, reason: "Competitive pressure from local providers" },
  apac: { delta: 2, reason: "Market-entry pricing program active" },
  latam: { delta: 3, reason: "FX exposure offset + growth incentive" },
};

export interface QuoteInput {
  product: ServerProduct;
  segment: CustomerSegment;
  units: number;
  termMonths: 12 | 24 | 36;
}

export function calculateDiscount({
  product,
  segment,
  units,
  termMonths,
}: QuoteInput): DiscountQuote {
  const rules: DiscountQuote["rules"] = [];

  const segmentRule = SEGMENT_RULES[segment];
  rules.push({
    label: `Segment baseline — ${segment.replace("-", " ")}`,
    delta: segmentRule.base,
    detail: segmentRule.note,
  });

  const tier = [...product.discountTiers]
    .filter((t) => units >= t.minUnits && t.label.startsWith(String(termMonths)))
    .sort((a, b) => b.percent - a.percent)[0];

  if (tier) {
    rules.push({
      label: `Commit tier — ${tier.label}`,
      delta: tier.percent,
      detail: `Published tier for ${tier.minUnits}+ units on this SKU`,
    });
  } else {
    const nearest = [...product.discountTiers].sort((a, b) => a.minUnits - b.minUnits)[0];
    rules.push({
      label: "Commit tier — not reached",
      delta: 0,
      detail: nearest
        ? `Needs ${nearest.minUnits} units on a ${nearest.label} to unlock ${nearest.percent}%`
        : "No published commit tiers for this SKU",
    });
  }

  const region = REGION_ADJUST[product.region]!;
  rules.push({
    label: `Region adjustment — ${REGION_SHORT[product.region]}`,
    delta: region.delta,
    detail: region.reason,
  });

  if (product.availability === "constrained" || product.availability === "backorder") {
    const penalty = product.availability === "backorder" ? -4 : -2;
    rules.push({
      label: "Supply constraint hold-back",
      delta: penalty,
      detail: `${product.leadTimeDays}-day lead time — pricing desk restricts concessions`,
    });
  }

  if (units >= 50) {
    rules.push({
      label: "Volume override",
      delta: 3,
      detail: `${units} units clears the 50-unit strategic volume gate`,
    });
  }

  const raw = rules.reduce((sum, r) => sum + r.delta, 0);
  const maxAllowed = 100 - product.marginFloorPercent;
  const discountPercent = Math.max(0, Math.min(raw, maxAllowed));

  if (raw > maxAllowed) {
    rules.push({
      label: "Margin floor cap",
      delta: +(maxAllowed - raw).toFixed(1),
      detail: `Capped at ${maxAllowed}% to protect the ${product.marginFloorPercent}% margin floor`,
    });
  }

  const listTotal = product.basePrice * units * termMonths;
  const netUnitPrice = product.basePrice * (1 - discountPercent / 100);
  const netTotal = netUnitPrice * units * termMonths;

  // Confidence reflects historical sample size, supply certainty and how
  // close the quote sits to the margin floor. Mirrors backend/app/discount.py —
  // each signal that moves the score records why, so the UI never has to guess.
  let score = 55;
  const confidenceFactors: DiscountQuote["confidenceFactors"] = [];
  const segmentText = segment.replace("-", " ");

  if (segmentRule.sampleSize > 200) {
    score += 20;
    confidenceFactors.push({
      label: "Deep deal history",
      detail: `${segmentRule.sampleSize} closed ${segmentText} deals back this rate`,
      direction: "up",
    });
  } else if (segmentRule.sampleSize > 100) {
    score += 12;
    confidenceFactors.push({
      label: "Moderate deal history",
      detail: `${segmentRule.sampleSize} comparable ${segmentText} deals on record`,
      direction: "up",
    });
  } else {
    score += 4;
    confidenceFactors.push({
      label: "Thin deal history",
      detail: `Only ${segmentRule.sampleSize} comparable ${segmentText} deals to reference`,
      direction: "down",
    });
  }

  if (tier) {
    score += 14;
    confidenceFactors.push({
      label: "Exact rate card match",
      detail: `Hits the published ${tier.label} tier at ${tier.minUnits}+ units`,
      direction: "up",
    });
  } else {
    confidenceFactors.push({
      label: "No commit tier matched",
      detail: "Priced off segment and region rules alone — no published tier applies",
      direction: "down",
    });
  }

  if (product.availability === "in-stock") {
    score += 10;
    confidenceFactors.push({
      label: "Supply confirmed",
      detail: `In stock, ${product.leadTimeDays}-day lead time`,
      direction: "up",
    });
  } else if (product.availability === "backorder") {
    score -= 12;
    confidenceFactors.push({
      label: "On backorder",
      detail: `${product.leadTimeDays}-day lead time puts the quote date at risk`,
      direction: "down",
    });
  } else {
    confidenceFactors.push({
      label: "Supply constrained",
      detail: `${product.leadTimeDays}-day lead time — allocation not guaranteed`,
      direction: "down",
    });
  }

  const marginPercent = 100 - product.marginFloorPercent - discountPercent;
  if (marginPercent < 5) {
    score -= 18;
    confidenceFactors.push({
      label: "Close to margin floor",
      detail: `Only ${marginPercent.toFixed(1)}% headroom above the ${product.marginFloorPercent}% floor`,
      direction: "down",
    });
  }

  score = Math.max(12, Math.min(97, score));

  const confidence = score >= 78 ? "high" : score >= 55 ? "medium" : "low";

  return {
    productId: product.id,
    segment,
    units,
    termMonths,
    basePrice: product.basePrice,
    listTotal,
    discountPercent: +discountPercent.toFixed(1),
    netUnitPrice: +netUnitPrice.toFixed(2),
    netTotal: +netTotal.toFixed(2),
    savings: +(listTotal - netTotal).toFixed(2),
    confidence,
    confidenceScore: score,
    marginPercent: +marginPercent.toFixed(1),
    requiresApproval: discountPercent > 20 || marginPercent < 6,
    rules,
    confidenceFactors,
  };
}

export const currency = (n: number, opts: Intl.NumberFormatOptions = {}) =>
  new Intl.NumberFormat("en-US", {
    style: "currency",
    currency: "USD",
    maximumFractionDigits: 0,
    ...opts,
  }).format(n);
